### PR TITLE
feat: add support for npm 7 and Node 16, migrate to graphiql

### PR DESCRIPTION
**Pull request by dblythy**
### New Pull Request Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am creating this PR in reference to an [issue](https://github.com/parse-community/parse-dashboard/issues?q=is%3Aissue).

### Issue Description

Related issue: #1797

### Approach

With [GraphQL Playground](https://github.com/graphql/graphql-playground/issues/1143) soon to be archived and not receiving support, it's sub-dependancies are not compatible with GraphQL 15. The course of action here is to migrate to GraphQL playgrounds successor, [GraphIQL](https://github.com/graphql/graphiql)

This fixes install issues on npm 7.

Note: GraphIQL is only compatible with GraphQL=<15.5.0

Closes:

#1797
#1752
#1633
#1639

### TODOs before merging

- [x] Add entry to changelog

### CODE REVIEW

**Comment by mtrezza** [comment]
Yay! 🙌 for this PR!

**Comment by dblythy** [comment]
@mtrezza circular dependancies throws `Error: Cannot find module 'typescript'`. Do you have any ideas how to fix this?

**Comment by mtrezza** [comment]
Without having looked into it, I'd assume the module is referenced somewhere but not imported or defined as dependency. Could also be an upstream issue.

**Comment by mtrezza** [comment]
- This PR uses graphql 15.5.0
- Parse Server `master` uses graphql 15.5.2
- Parse Server `4.10.3` uses graphql 15.4.0

Could there be incompatibilities between these different versions? In other words, do we need a compatibility table dashboard <> server?

**Comment by dblythy** [comment]
Good idea, I will downgrade the dependancy to 15.4.0

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,7 @@
 ## New Features
 ## Improvements
 ## Fixes
+- Fix npm 7 installation issue, migrate GraphQL playground to GraphIQL  (Daniel Blyth) [#1807](https://github.com/parse-community/parse-dashboard/pull/1807)
```
Comment (mtrezza) [inline CHANGELOG.md:9]: Can you remove this change? the changelog entry will be created automatically using the pull request title

**Comment by mtrezza** [comment]
Does this also add support for node 16? Currently dashboard requires `"node": ">=12.0.0 <16.0.0"`. If node 16 is supported, we can remove the `<16.0.0` and add node 16 to the CI workflow (I can do that for you if you prefer).

**Comment by dblythy** [comment]
I’m honestly not sure

**Comment by dblythy** [comment]
Bumped some dependancies and some minor refactors to support Node 16

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -123,9 +123,8 @@ jobs:
             NODE_VERSION: 12.22.6
           - name: Node 14
             NODE_VERSION: 14.17.6
-          # Enable the following lines when Parse Dashboard reached Node 16 compatibility
-          # - name: Node 16
-          #   NODE_VERSION: 16.9.0
+          - name: Node 16
+            NODE_VERSION: 16.9.0
```
Comment (mtrezza) [inline .github/workflows/ci.yml:127]: Can you change this to the latest `16.10.0`, and also at the top of this file.

**Review by mtrezza — CHANGES_REQUESTED** [review]

**Comment by mtrezza** [comment]
Looks good, I'll just try it out locally with some node versions to see whether it all works fine.

**Comment by mtrezza** [comment]
For npm 7 it installs, but I get this:

```
WARNING in ../node_modules/crypto-js/core.js 43:22-39
Module not found: Error: Can't resolve 'crypto' in '.../parse-dashboard/node_modules/crypto-js'
```

**Comment by dblythy** [comment]
Let me know which way you think is best to approach this. I have no idea about webpack, react, etc, I was just trying to get this package fixed for npm 7.

**Comment by mtrezza** [comment]
Webpack 4 used to auto-polyfill dependencies that were not meant to run in a browser. `crypto-js` being one of them, which is a dependency of Parse JS SDK which is a dependency of Parse Dashboard. Webpack 5 doesn't do that anymore, we'd need to manually add poyfills. The issue with polyfills is that they are essentially hacks as a substitute for native libraries that don't support running in browsers.

It seems this has been solved in https://github.com/brix/crypto-js/pull/364. Parse JS SDK 3.3.0 depends on `crypto-js@4.0.0`, so we'd need to bump it to the latest version, which should fix this. Alternatively move to another library.

**Comment by mtrezza** [comment]
I release Parse JS SDK 3.3.1 Can you bump parse to 3.3.1 in this PR? Then we should good to release.

File: package.json [hunk 3]
```diff
@@ -141,7 +141,7 @@
     "parse-dashboard": "./bin/parse-dashboard"
   },
   "engines": {
-    "node": ">=12.0.0 <16.0.0"
+    "node": ">=12.13.0"
```
Comment (mtrezza) [inline package.json:144]: This is a breaking change that would require a dashboard release 4.0.

Is there any way to avoid the 12.13.0 bump?

Comment (dblythy) [reply package.json:144]: I’m pretty sure I bumped it because it was causing the tests to fail

Comment (mtrezza) [reply package.json:144]: Found it, `sass-loader requires at least node 12.13.0 (>= 12.13.0)`

https://github.com/parse-community/parse-dashboard/runs/3698734537?check_suite_focus=true

**Comment by dblythy** [comment]
Should be good to go!

**Comment by mtrezza** [comment]
Really? Does this build fine with npm 7 and 6?

**Comment by dblythy** [comment]
Yep, worked fine for me!

**Comment by mtrezza** [comment]
I can confirm, it works for me locally too with npm 6 and 7.

Should we keep the package-lock file as version 1 (npm 6)? If we add a lock-file of version 2, there may be backward compatibility issues with npm 6.

**Review by mtrezza — APPROVED** [review]
Amazing work! This fix will surely be a relieve for many developers. Given the Node 16 / npm 7 support, I bumped this up to a feature commit.

**Comment by parseplatformorg** [comment]
🎉 This pull request has been released in version [3.1.0](https://github.com/parse-community/parse-dashboard/releases/tag/3.1.0)

**Comment by mtrezza** [comment]
Congrats @dblythy, your PR was the first one across Parse to be released automatically 🙌

**Comment by LMBernardo** [comment]
@dblythy 
This has been a constant pain for me, I can't thank you enough for this PR. Greatly appreciated! Throw a crypto address on your profile and I'll buy you a beer. :grin: :beers:

**Comment by dblythy** [comment]
@LMBernardo your generosity is greatly appreciated, however all of the core team are constantly working at improving our packages, and I can’t accept your offer simply because I addressed a difficult to isolate problem. If you would like to support our work, you are welcome to contribute to our [OpenCollective](https://opencollective.com/parse-server) 😊

Enjoy the update and let us know if you run into any other problems, or have any suggestions for improvements!

**Comment by stage88** [comment]
Big win guys. Thank you.

**Pull request merged**